saleor/graphql/discount/mutations.py

```diff
@@ -0,0 +1,154 @@
+import graphene
+
+from ...discount import VoucherType, models
+from ..core.mutations import ModelDeleteMutation, ModelMutation
+from ..core.types import Decimal
+
+
+def validate_voucher(voucher_data):
+    voucher_type = voucher_data.get('type')
+    errors = []
+    if voucher_type == VoucherType.PRODUCT:
+        if not voucher_data.get('product'):
+            errors.append(('product', 'This field is required.'))
+    elif voucher_type == VoucherType.CATEGORY:
+        if not voucher_data.get('category'):
+            errors.append(('category', 'This field is required.'))
+    return errors
+
+
+class VoucherInput(graphene.InputObjectType):
+    type = graphene.String(
+        description='Voucher type: product, category shipping or value.')
+    name = graphene.String(description='Voucher name.')
+    code = graphene.String(decription='Code to use the voucher.')
+    start_date = graphene.types.datetime.DateTime(
+        description='Start date of the voucher in ISO 8601 format.')
+    end_date = graphene.types.datetime.DateTime(
+        description='End date of the voucher in ISO 8601 format.')
+    discount_value_type = graphene.String(
+        description='Choices: fixed or percentage.')
+    discount_value = Decimal(description='Value of the voucher.')
+    product = graphene.ID(description='Product related to the discount.')
+    category = graphene.ID(description='Category related to the discount.')
+    apply_to = graphene.String(
+        description='Single item (one) or all matching products (all).')
+    limit = Decimal(description='Limit value of the discount.')
+
+
+class VoucherCreate(ModelMutation):
+    class Arguments:
+        input = VoucherInput(
+            required=True,
+            description='Fields required to create a voucher.')
+
+    class Meta:
+        description = 'Creates a new voucher.'
+        model = models.Voucher
+
+    @classmethod
+    def user_is_allowed(cls, user, input):
+        return user.has_perm('discount.edit_voucher')
+
+    @classmethod
+    def clean_input(cls, info, instance, input, errors):
+        cleaned_input = super().clean_input(info, instance, input, errors)
+        voucher_errors = validate_voucher(cleaned_input)
+        for err in voucher_errors:
+            cls.add_error(errors=errors, field=err[0], message=err[1])
+        return cleaned_input
+
+
+class VoucherUpdate(ModelMutation):
```

Comment:
We could use inheritance here and extend the `VoucherCreate` to save a few lines which are the same at the moment in both mutations.